### PR TITLE
DSET-680 Extend agent status

### PR DESCRIPTION
**Pull request by ArthurKamalov**
This PR adds new agent stat value which represents average time that is needed to generate and write agent status.
This new stat can be seen in agent's status and in the debug log.

Agent status report also has been modified to print some useful data about system if status request has ended with timeout.

Other fixes:
- Change initialisation logic of the k8s cache. Now it is created during the first access instead of  being created directly in global scope, which causes log messages leak to agent status.

here's the sample output 

```
scalyr-agent-2 % python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1
Failed to get status within 5 seconds.  Giving up.  The agent process is possibly stuck.  See /Users/arthur/scalyr-agent-dev/log/agent.log for more details.
Debug stats: [
    {
        "agent_process": {
            "cpu_percent": 0.0,
            "memory_rss": 49733632,
            "pid": 14652
        },
        "cpu_count": 12,
        "cpu_percent": 15.4,
        "getloadavg": [
            2.1591796875,
            2.765625,
            2.84033203125
        ],
        "processes": [
            "arthur           14684  56.6  0.1 34920852  39520 s002  S+    4:36PM   0:00.29 python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1",
            "arthur           14652   0.4  0.1 35033600  48568   ??  S     4:36PM   0:00.54 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Users/arthur/work/agents/scalyr-agent-2/scalyr_agent/agent_main.py start --no-fork",
            "arthur           62852   0.1  0.0 35381940  31276   ??  S    19Oct22  34:30.11 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py --mode=client --port=54569"
        ],
        "status_file_info": {
            "content": "",
            "data_root_file_stats": [
                "total 96",
                "drwxr-xr-x  12 arthur  staff   384 Nov  9 16:36 .",
                "drwxr-xr-x   5 arthur  staff   160 Aug 24 11:26 ..",
                "-rw-r--r--@  1 arthur  staff  6148 Nov  8 16:07 .DS_Store",
                "-rw-r--r--   1 arthur  staff   356 Nov  9 16:36 active-checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 active-checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff  2457 Nov  9 16:36 checkpoints.json",
                "-rw-r--r--   1 arthur  staff     0 Nov  9 16:36 last_status",
                "-rw-------   1 arthur  staff  8161 Nov  8 15:28 last_status_h6m8308e.tmp",
                "-rw-------   1 arthur  staff  8164 Nov  8 15:28 last_status_pthf0f98.tmp",
                "-rw-r--r--@  1 arthur  staff     4 Nov  9 16:36 status_format"
            ],
            "temp_status_files_stats": "name 'TEMP_STATUS_FILES_DIR_NAME' is not defined"
        },
        "timestamp": 216908000,
        "virtual_memory_percent": 48.7
    },
    {
        "agent_process": {
            "cpu_percent": 0.0,
            "memory_rss": 49737728,
            "pid": 14652
        },
        "cpu_count": 12,
        "cpu_percent": 15.8,
        "getloadavg": [
            2.1591796875,
            2.765625,
            2.84033203125
        ],
        "processes": [
            "arthur           14684  54.7  0.1 35051924  40036 s002  S+    4:36PM   0:00.30 python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1",
            "arthur           14652   0.2  0.1 35033600  48572   ??  S     4:36PM   0:00.54 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Users/arthur/work/agents/scalyr-agent-2/scalyr_agent/agent_main.py start --no-fork",
            "arthur           62852   0.1  0.0 35381940  31276   ??  S    19Oct22  34:30.11 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py --mode=client --port=54569"
        ],
        "status_file_info": {
            "content": "",
            "data_root_file_stats": [
                "total 96",
                "drwxr-xr-x  12 arthur  staff   384 Nov  9 16:36 .",
                "drwxr-xr-x   5 arthur  staff   160 Aug 24 11:26 ..",
                "-rw-r--r--@  1 arthur  staff  6148 Nov  8 16:07 .DS_Store",
                "-rw-r--r--   1 arthur  staff   356 Nov  9 16:36 active-checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 active-checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff  2457 Nov  9 16:36 checkpoints.json",
                "-rw-r--r--   1 arthur  staff     0 Nov  9 16:36 last_status",
                "-rw-------   1 arthur  staff  8161 Nov  8 15:28 last_status_h6m8308e.tmp",
                "-rw-------   1 arthur  staff  8164 Nov  8 15:28 last_status_pthf0f98.tmp",
                "-rw-r--r--@  1 arthur  staff     4 Nov  9 16:36 status_format"
            ],
            "temp_status_files_stats": "name 'TEMP_STATUS_FILES_DIR_NAME' is not defined"
        },
        "timestamp": 387674000,
        "virtual_memory_percent": 48.7
    },
    {
        "agent_process": {
            "cpu_percent": 0.0,
            "memory_rss": 49737728,
            "pid": 14652
        },
        "cpu_count": 12,
        "cpu_percent": 9.5,
        "getloadavg": [
            2.1591796875,
            2.765625,
            2.84033203125
        ],
        "processes": [
            "arthur           14684   0.9  0.1 35051924  40240 s002  S+    4:36PM   0:00.31 python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1",
            "arthur           62852   0.1  0.0 35381940  31276   ??  S    19Oct22  34:30.11 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py --mode=client --port=54569",
            "arthur           14652   0.0  0.1 35033600  48572   ??  S     4:36PM   0:00.54 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Users/arthur/work/agents/scalyr-agent-2/scalyr_agent/agent_main.py start --no-fork"
        ],
        "status_file_info": {
            "content": "",
            "data_root_file_stats": [
                "total 96",
                "drwxr-xr-x  12 arthur  staff   384 Nov  9 16:36 .",
                "drwxr-xr-x   5 arthur  staff   160 Aug 24 11:26 ..",
                "-rw-r--r--@  1 arthur  staff  6148 Nov  8 16:07 .DS_Store",
                "-rw-r--r--   1 arthur  staff   356 Nov  9 16:36 active-checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 active-checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff  2457 Nov  9 16:36 checkpoints.json",
                "-rw-r--r--   1 arthur  staff     0 Nov  9 16:36 last_status",
                "-rw-------   1 arthur  staff  8161 Nov  8 15:28 last_status_h6m8308e.tmp",
                "-rw-------   1 arthur  staff  8164 Nov  8 15:28 last_status_pthf0f98.tmp",
                "-rw-r--r--@  1 arthur  staff     4 Nov  9 16:36 status_format"
            ],
            "temp_status_files_stats": "name 'TEMP_STATUS_FILES_DIR_NAME' is not defined"
        },
        "timestamp": 627134000,
        "virtual_memory_percent": 48.8
    },
    {
        "agent_process": {
            "cpu_percent": 0.0,
            "memory_rss": 49737728,
            "pid": 14652
        },
        "cpu_count": 12,
        "cpu_percent": 6.2,
        "getloadavg": [
            2.1591796875,
            2.765625,
            2.84033203125
        ],
        "processes": [
            "arthur           14684   0.3  0.1 35051924  40336 s002  S+    4:36PM   0:00.32 python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1",
            "arthur           62852   0.1  0.0 35381940  31276   ??  S    19Oct22  34:30.11 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py --mode=client --port=54569",
            "arthur           14652   0.0  0.1 35033600  48572   ??  S     4:36PM   0:00.54 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Users/arthur/work/agents/scalyr-agent-2/scalyr_agent/agent_main.py start --no-fork"
        ],
        "status_file_info": {
            "content": "",
            "data_root_file_stats": [
                "total 96",
                "drwxr-xr-x  12 arthur  staff   384 Nov  9 16:36 .",
                "drwxr-xr-x   5 arthur  staff   160 Aug 24 11:26 ..",
                "-rw-r--r--@  1 arthur  staff  6148 Nov  8 16:07 .DS_Store",
                "-rw-r--r--   1 arthur  staff   356 Nov  9 16:36 active-checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 active-checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff  2457 Nov  9 16:36 checkpoints.json",
                "-rw-r--r--   1 arthur  staff     0 Nov  9 16:36 last_status",
                "-rw-------   1 arthur  staff  8161 Nov  8 15:28 last_status_h6m8308e.tmp",
                "-rw-------   1 arthur  staff  8164 Nov  8 15:28 last_status_pthf0f98.tmp",
                "-rw-r--r--@  1 arthur  staff     4 Nov  9 16:36 status_format"
            ],
            "temp_status_files_stats": "name 'TEMP_STATUS_FILES_DIR_NAME' is not defined"
        },
        "timestamp": 827924000,
        "virtual_memory_percent": 48.7
    },
    {
        "agent_process": {
            "cpu_percent": 0.0,
            "memory_rss": 49737728,
            "pid": 14652
        },
        "cpu_count": 12,
        "cpu_percent": 6.8,
        "getloadavg": [
            2.14599609375,
            2.75244140625,
            2.8349609375
        ],
        "processes": [
            "arthur           14684   0.3  0.1 35051924  40512 s002  S+    4:36PM   0:00.33 python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1",
            "arthur           62852   0.1  0.0 35381940  31276   ??  S    19Oct22  34:30.12 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py --mode=client --port=54569",
            "arthur           14652   0.0  0.1 35033600  48572   ??  S     4:36PM   0:00.54 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Users/arthur/work/agents/scalyr-agent-2/scalyr_agent/agent_main.py start --no-fork"
        ],
        "status_file_info": {
            "content": "",
            "data_root_file_stats": [
                "total 96",
                "drwxr-xr-x  12 arthur  staff   384 Nov  9 16:36 .",
                "drwxr-xr-x   5 arthur  staff   160 Aug 24 11:26 ..",
                "-rw-r--r--@  1 arthur  staff  6148 Nov  8 16:07 .DS_Store",
                "-rw-r--r--   1 arthur  staff   356 Nov  9 16:36 active-checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 active-checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff  2457 Nov  9 16:36 checkpoints.json",
                "-rw-r--r--   1 arthur  staff     0 Nov  9 16:36 last_status",
                "-rw-------   1 arthur  staff  8161 Nov  8 15:28 last_status_h6m8308e.tmp",
                "-rw-------   1 arthur  staff  8164 Nov  8 15:28 last_status_pthf0f98.tmp",
                "-rw-r--r--@  1 arthur  staff     4 Nov  9 16:36 status_format"
            ],
            "temp_status_files_stats": "name 'TEMP_STATUS_FILES_DIR_NAME' is not defined"
        },
        "timestamp": 45703000,
        "virtual_memory_percent": 48.7
    },
    {
        "agent_process": {
            "cpu_percent": 0.0,
            "memory_rss": 49750016,
            "pid": 14652
        },
        "cpu_count": 12,
        "cpu_percent": 6.9,
        "getloadavg": [
            2.14599609375,
            2.75244140625,
            2.8349609375
        ],
        "processes": [
            "arthur           14652   0.8  0.1 35033600  48584   ??  S     4:36PM   0:00.55 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Users/arthur/work/agents/scalyr-agent-2/scalyr_agent/agent_main.py start --no-fork",
            "arthur           14684   0.3  0.1 35051924  40552 s002  S+    4:36PM   0:00.35 python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1",
            "arthur           62852   0.1  0.0 35381940  31276   ??  S    19Oct22  34:30.12 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py --mode=client --port=54569"
        ],
        "status_file_info": {
            "content": "",
            "data_root_file_stats": [
                "total 96",
                "drwxr-xr-x  12 arthur  staff   384 Nov  9 16:36 .",
                "drwxr-xr-x   5 arthur  staff   160 Aug 24 11:26 ..",
                "-rw-r--r--@  1 arthur  staff  6148 Nov  8 16:07 .DS_Store",
                "-rw-r--r--   1 arthur  staff   356 Nov  9 16:36 active-checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 active-checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff  2458 Nov  9 16:36 checkpoints.json",
                "-rw-r--r--   1 arthur  staff     0 Nov  9 16:36 last_status",
                "-rw-------   1 arthur  staff  8161 Nov  8 15:28 last_status_h6m8308e.tmp",
                "-rw-------   1 arthur  staff  8164 Nov  8 15:28 last_status_pthf0f98.tmp",
                "-rw-r--r--@  1 arthur  staff     4 Nov  9 16:36 status_format"
            ],
            "temp_status_files_stats": "name 'TEMP_STATUS_FILES_DIR_NAME' is not defined"
        },
        "timestamp": 227143000,
        "virtual_memory_percent": 48.7
    },
    {
        "agent_process": {
            "cpu_percent": 0.0,
            "memory_rss": 49750016,
            "pid": 14652
        },
        "cpu_count": 12,
        "cpu_percent": 16.3,
        "getloadavg": [
            2.14599609375,
            2.75244140625,
            2.8349609375
        ],
        "processes": [
            "arthur           14684   1.6  0.1 35051924  40744 s002  S+    4:36PM   0:00.35 python3 scalyr_agent/agent_main.py status -H --debug --stats-capture-interval 1",
            "arthur           14652   0.7  0.1 35033600  48584   ??  S     4:36PM   0:00.55 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Users/arthur/work/agents/scalyr-agent-2/scalyr_agent/agent_main.py start --no-fork",
            "arthur           62852   0.1  0.0 35381940  31276   ??  S    19Oct22  34:30.12 /Users/arthur/miniconda3/envs/scalyr-agent-2/38/bin/python /Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py --mode=client --port=54569"
        ],
        "status_file_info": {
            "content": "",
            "data_root_file_stats": [
                "total 96",
                "drwxr-xr-x  12 arthur  staff   384 Nov  9 16:36 .",
                "drwxr-xr-x   5 arthur  staff   160 Aug 24 11:26 ..",
                "-rw-r--r--@  1 arthur  staff  6148 Nov  8 16:07 .DS_Store",
                "-rw-r--r--   1 arthur  staff   356 Nov  9 16:36 active-checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 active-checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff    46 Nov  9 16:36 checkpoints-worker-default-0.json",
                "-rw-r--r--   1 arthur  staff    47 Nov  9 16:36 checkpoints-worker-wei-0.json",
                "-rw-r--r--   1 arthur  staff  2458 Nov  9 16:36 checkpoints.json",
                "-rw-r--r--   1 arthur  staff     0 Nov  9 16:36 last_status",
                "-rw-------   1 arthur  staff  8161 Nov  8 15:28 last_status_h6m8308e.tmp",
                "-rw-------   1 arthur  staff  8164 Nov  8 15:28 last_status_pthf0f98.tmp",
                "-rw-r--r--@  1 arthur  staff     4 Nov  9 16:36 status_format"
            ],
            "temp_status_files_stats": "name 'TEMP_STATUS_FILES_DIR_NAME' is not defined"
        },
        "timestamp": 409470000,
        "virtual_memory_percent": 48.7
    }
]

```

### CODE REVIEW

File: scalyr_agent/agent_main.py [hunk 1]
```diff
@@ -689,6 +700,40 @@ def __handle_terminate(self):
...
+            raise Exception("Pid file does not exist.")
+
+        with open(pidfile, "r") as fp:
+            pidfile_content = fp.read().strip()
+
+        process = psutil.Process(
+            pid=int(pidfile_content)
+        )
+
+        return {
```
Comment (Kami) [inline scalyr_agent/agent_main.py:723]: Per slack discussion the other day, I'm slightly worried that this logging will not be sufficient to track down the root cause.

I propose adding this additional logging:

* Use psutil or similar to run command similar to "ps aux | grep scalyr-agent-2" and log the output. This may help us catch an issue if there is a stuck process from a previous run or similar.
* This information should be logged multiple times - before executing status, during (while sleeping and waiting) and at the end. That's important because it's likely a race condition or similar and printing it just at the end likely won't be sufficient.

**Comment by ArthurKamalov** [comment]
@Kami Was trying to solve problems with Windows, but it seems now everything is OK. 

Now health check status can be requested with the `--debug` flag that will print additional info. 
This info now also includes agent process info, all agent related processes list and some machine stats. All this info is captured at the beginning, during waiting and at the end.

File: scalyr_agent/agent_main.py [hunk 2]
```diff
@@ -689,12 +706,93 @@ def __handle_terminate(self):
...
+                machine_stats[
+                    "virtual_memory_percent"
+                ] = psutil.virtual_memory().percent
+            except Exception as e:
+                machine_stats["error"] = str(e)
+
+            return machine_stats
+
+        # Get list of related processes'
+        def find_agent_processed():
```
Comment (Kami) [inline scalyr_agent/agent_main.py:762]: processed -> processes

File: scalyr_agent/agent_main.py [hunk 3]
```diff
@@ -794,8 +909,15 @@ def __detailed_status(
         # We wait for five seconds at most to get the status.
         deadline = time.time() + 5
 
+        last_debug_stat_time = 0
         # Now loop until we see it show up.
         while True:
+            # Capture debug stats every second while waiting.
+            if debug and (last_debug_stat_time + 1 < time.time()):
```
Comment (Kami) [inline scalyr_agent/agent_main.py:916]: I wonder if we should make that value configurable via command line options because in case there is indeed a race condition, we may need to use lower value (e.g. 0.5 seconds or less) to try to find offending stray process which may be causing a race condition.

File: scalyr_agent/agent_main.py [hunk 4]
```diff
@@ -716,6 +815,22 @@ def __detailed_status(
         if health_check:
             status_format = "json"
 
+        # List of all debug stats that are captured during status report.
+        debug_stats = []
+
+        def print_debug_stats():
+            print(
```
Comment (Kami) [inline scalyr_agent/agent_main.py:822]: Wonder if we should take ``status_format`` command line option value into account here as well and only use machine readable json format in case json is used, otherwise default to more user-friendly format.

WDYT? Or is this an over kill / not needed?

File: scalyr_agent/agent_main.py (hunk 3 above)
Comment (Kami) [reply scalyr_agent/agent_main.py:916]: EDIT: Never mind, I noticed below it only works with health check health option right now so this change may not be needed.

File: scalyr_agent/agent_main.py [hunk 5]
```diff
@@ -689,12 +706,93 @@ def __handle_terminate(self):
...
+            except Exception as e:
+                return {"error": "Can't get list of processes. Error: {}".format(e)}
+
+            for line in ps_output.splitlines():
+                if "scalyr-agent" in line or "python" in line or "agent_main" in line:
+                    agent_processes.append(line)
+
+            return agent_processes
+
+        result = {}
```
Comment (Kami) [inline scalyr_agent/agent_main.py:780]: I wonder if it makes sense to also try to include the output of equivalent of ``ls -la status_file`` and ``cat status_file`` here.

In case there is some race condition with writing / reading a file, this may make it easier to catch it.

File: scalyr_agent/agent_main.py (hunk 4 above)
Comment (ArthurKamalov) [reply scalyr_agent/agent_main.py:844]: I am not sure that it can be more user-friendly with the amount of information that we gather here. I think it has to be readable at least at some level with pretty-printing the json, that we already do.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,17 @@
 Scalyr Agent 2 Changes By Release
 =================================
 
+## 2.1.38 "TBD" - October 17, 2022
+<!---
+Packaged by Joseph Makar <arthurk@sentinelone.com> on Oct 17, 2022 12:31 -0800
+--->
+
+Kubernetes:
+* Fix bug that caused logging of the Kubernetes cache stats to agent status.
```
Comment (Kami) [inline CHANGELOG.md:10]: Good catch, thanks. 

We had a bunch of similar issues in the past (logs leaking status output which can break json parsing). We do have tests for it, but it's almost impossible to test all scenarios since if some log message is logged can depend on config option and environment (we would need to test all the possible permutations).

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,17 @@
...
+## 2.1.38 "TBD" - October 17, 2022
+<!---
+Packaged by Joseph Makar <arthurk@sentinelone.com> on Oct 17, 2022 12:31 -0800
+--->
+
+Kubernetes:
+* Fix bug that caused logging of the Kubernetes cache stats to agent status. 
+
+Other
+* Add to agent's status its generation and average generation time.
```
Comment (Kami) [inline CHANGELOG.md:13]: Might want to reword it to say something along the line of "Added support for --debug flag to the scalyr-agent-2 status command. When this flag is used, agent prints additional debug related information with the status output. NOTE: Right now it's only supported with the healthcheck config option".

File: scalyr_agent/agent_main.py [hunk 8]
```diff
@@ -689,12 +706,112 @@ def __handle_terminate(self):
...
+                    pidfile_content = fp.read().strip()
+            except OSError as e:
+                return {"error": "Error during PID file read. Error: {}".format(e)}
+
+            try:
+                psutil_process = psutil.Process(pid=int(pidfile_content))
+            except psutil.Error as e:
+                return {"error": "Can't create psutil process. Error: {}".format(e)}
+
+            try:
```
Comment (Kami) [inline scalyr_agent/agent_main.py:737]: Minor (performance) improvement - may want to use ``oneshot()`` context manager - https://psutil.readthedocs.io/en/latest/#psutil.Process.oneshot

Honestly, performance is not an issue here, but it may make sense to use oneshot in case there is a race (so we retrieve all the info / metrics for a running process at the same time - in case the process dies / goes away when we try to retrieve some other metric - probably not all that likely to happen though).

**Comment by Kami** [comment]
From the sample output:

```
        "agent_process": {
            "cpu_percent": 36007936
        },
```

This appears to be a counter (possible jiffies or us unit) and not percent. So we should probably clear this up (leave value as is, change name so it correctly indicates what kind of value it contains and / or add new cpu_percent metric which includes actual percentage value).

**Comment by Kami** [comment]
Also, let's please include unix timestamp in milliseconds with each status array item.

File: CHANGELOG.md (hunk 7 above)
Comment (ArthurKamalov) [reply CHANGELOG.md]: Thanks, copy-pasted it :)

File: scalyr_agent/agent_main.py [hunk 9]
```diff
@@ -794,8 +928,15 @@ def __detailed_status(
         # We wait for five seconds at most to get the status.
         deadline = time.time() + 5
 
+        last_debug_stat_time = 0
         # Now loop until we see it show up.
         while True:
+            # Capture debug stats every second while waiting.
+            if debug and (last_debug_stat_time + 1 < time.time()):
+                stats = self._get_system_and_agent_stats(status_file_path=status_file)
```
Comment (Kami) [inline scalyr_agent/agent_main.py:936]: I wonder if it would make sense to also try to gather output from last_status.tmp file to see if there is a race and something else is also requesting output to be written to that file (temporary file we write status to before we move it in an atomic fashion to a final location).

**Comment by ArthurKamalov** [comment]
> From the sample output:
> 
> ```
>         "agent_process": {
>             "cpu_percent": 36007936
>         },
> ```
> 
> This appears to be a counter (possible jiffies or us unit) and not percent. So we should probably clear this up (leave value as is, change name so it correctly indicates what kind of value it contains and / or add new cpu_percent metric which includes actual percentage value).

oops, I've messed up with dict keys here, thanks for the catch.

File: scalyr_agent/agent_main.py [hunk 10]
```diff
@@ -794,8 +928,15 @@ def __detailed_status(
         # We wait for five seconds at most to get the status.
         deadline = time.time() + 5
 
+        last_debug_stat_time = 0
         # Now loop until we see it show up.
         while True:
+            # Capture debug stats every second while waiting.
+            if debug and (last_debug_stat_time + 1 < time.time()):
```
Comment (Kami) [inline scalyr_agent/agent_main.py:935]: Per my comment above, I think it may be a good idea to allow this interval to be overridden / changed via command line flag (e.g. --stats-capture-interval 0.5 or similar).

File: scalyr_agent/agent_main.py (hunk 9 above)
Comment (ArthurKamalov) [reply scalyr_agent/agent_main.py]: I added creation of the `last_status.tmp` file by using the `mkstemp` as you've suggested in slack, so I guess, it may be enough just so see how many temp files now exist in the directory (by extending existing `ls` command) to see if there are multiple status requests at the same time. WDYT?

File: scalyr_agent/agent_main.py [hunk 11]
```diff
@@ -2256,18 +2275,33 @@ def __report_status_to_file(self):
             status_format = "text"
 
         tmp_file = None
+
         try:
+
+            # Before writing status file, search and delete status files from the previous status reports, which
+            # may be left undeleted because of unexpected errors. Since the `mkstemp` function creates uniquely
+            # named file, the number of such undeleted files may eventually grow and pollute the directory.
+            fount_tmp_files = list(glob.glob(
```
[the remark this answers is not in the document]
Comment (Kami) [reply scalyr_agent/agent_main.py]: And this delete itself may result / contribute to the race, so we should probably remove it all together to remove additional issues.

**Comment by ArthurKamalov** [comment]
@Kami I've rolled back `mkstemp` changes and also updated sample status output  in the description.

File: scalyr_agent/agent_main.py [hunk 12]
```diff
@@ -689,9 +708,126 @@ def __handle_terminate(self):
...
+                result["data_root_file_stats"] = ls_output.splitlines()
+            except Exception as e:
+                result["data_root_file_stats"] = six.text_type(e)
+
+            try:
+                ls_output = subprocess.check_output(
+                    [
+                        "ls",
+                        "-la",
+                        os.path.join(tempfile.gettempdir(), TEMP_STATUS_FILES_DIR_NAME),
```
Comment (Kami) [inline scalyr_agent/agent_main.py:801]: Did you forgot to update / revert this place back to using agent data dir?

Comment (ArthurKamalov) [reply scalyr_agent/agent_main.py]: ah, yes, removed. Thanks for noticing that.

**Review by Kami — APPROVED** [review]
